Realtime Implementation

rshar1 opened the pull request:
## Server Side: 

### New Class: BroadCastSystem

- Keep's track of which users are in which chat rooms
- Public methods that switch the chat room for each conversation and add messages to a queue of messages to be broadcasted
- A separate thread is used to send messages to users. This is done so users do not need to wait on server response when they add a message. Instead message will be added to queue and server will broadcast the message when it can. 

Since the message broadcaster has its own thread for sending out new messages, the server synchronizes the output stream whenever it sends a response. This is to ensure that server responses are not mixed with broadcasts. 

## Client Side:

### New Class: BroadCastReceiver

- The BroadCastReceiver is a thread on the client side
- Functional Interface BroadCastEvent will represent what the receiver does when a new message is pushed from the server. This was done so the BroadCastReceiver is independent of the user interface
- Listens to the port for any incoming connections from the server. If the server is pushing a new message, the BroadCastEvent will be called. 

When the client sends any requests to the server, such as join conversation or GetUsers, the requests must use the broadcastreceiver to wait for server response. The methods getType, and getInputStream are blocking and will only return when the server sends a response that isn't a broadcast. 

The view and controller have been changed so they no longer close the connection. Instead they call the broadcast receiver to get the input stream. 


## Common to both: 
Network codes have been added for new broadcasts and switching conversations.

rshar1 committed:
added network codes

added the new network codes to represent the broadcast changes to be
made

rshar1 committed:
created broadcast hub and sytem

Created initial broadcast hub and broadcast system. This will run along
side the original server socket but will only be sending out new
messages

rshar1 committed:
implemented methods for broadcast system

Implemented the methods for the broadcast system on the server side

rshar1 committed:
modified broadcast hub to make nonblocking

Modified the broadcast hub to make it nonblocking. This would allow the
hub to run.

Also, began implementation of handleCommand method in the broadcast
system

rshar1 committed:
implemented handle command

Implemented handle command so that user will now switch conversations
when they send join conversation request

rshar1 committed:
implemented connection listener thread

Implemented the connection listener thread. This will constantly listen
to the join requests from the client and then change the conversation
if requested

rshar1 committed:
removed close of broadCastSource

removed the close of the broadcast source and placed it in a try with
block.

rshar1 committed:
made some bug fixes

Just some bug fixes

rshar1 committed:
pushing now works

Conversation Uuid is now sent to broadcaster so that it can broadcast
the message.

rshar1 committed:
renamed class and bug fix

Renamed the BroadCastReceiver to correct a typo in the class name.

Also fixed a bug in which user couldn’t join conversation until they
first start a new conversation

rshar1 committed:
change for client to exit

Added exit method to the broadcast Receiver. This will end the thread
on the client side.

rshar1 committed:
added comments and slight changes

Added comments to the code to clarify some of the additions that have
been made.

Updated the port for the broadcast system so that it is now given as an
argument.

rshar1 committed:
Added broadcast receiver to simple gui

Added the broadcast receiver to the simple GUI.

This changes will serve as sample for how a gui application would work
in real time.

rshar1 committed:
fixed bug on server side

Fixed a bug on the server side which caused the listener thread to stay
open after the client was opened. This was caused because the
serializers currently return a value of -1 if a null value was read by
the input stream.

rshar1 committed:
Added comment for clarity

Added a clarification comment to the broadcast system.

rshar1 committed:
removed the broadcast hub

Removed the broadcast hub and began making changes to add the broadcast
system to the same port as the server

rshar1 committed:
synchronized output stream in server

Synchronized the output stream in the server to ensure that the
messages are sent back together.

rshar1 committed:
added response event interface

Added a new functional interface to the broadcast receiver. This
interface is called when a response arrives from the server. It will be
used to properly execute what must be done on each response

rshar1 committed:
broadcast reciever now handles responses

Added some additional methods to the broadcast receiver. Broad cast
receiver has some blocking methods that will wait for response before
returning.

rshar1 committed:
limited source control and fixed controller

Removed access of the connection source from the controller and the
view. All input and output must go through the broadcast receiver on
the client side. Also corrected a typo.

rshar1 committed:
implemented the view

Implemented the view to work using the same connection as the broadcast
receiver

rshar1 committed:
fixed issue in controller code

Fixed a bug in the code because of some code that I forgot to fix
earlier.

This is the first working version with realtime running on only one
port. (Only works for command line)

rshar1 committed:
realtime on OnePort working on simple gui

Fixed up the simple gui to work with the new view and controller. This
allows the simple gui to now accept new messages and responses on the
same port

rshar1 committed:
fixed else statement

Removed spacing before else if statement.

rshar1 committed:
minor cleanup

Removed unnecessary code and comments.

rshar1 committed:
Merge branch 'realtimeOnePort' into implementingRealtime

# Conflicts:
#	src/codeu/chat/client/BroadCastReceiver.java

rshar1 committed:
Merge remote-tracking branch 'origin/testing' into realtimeOnePort

# Conflicts:
#	src/codeu/chat/client/BroadCastReceiver.java

Commit:
Merge pull request #8 from turtlefarmer/implementingRealtime

Implementation of Real Time

rshar1 committed:
Merge branch 'realtimeOnePort' into implementingRealtime

rshar1 committed:
Merge pull request #13 from turtlefarmer/implementingRealtime

updating realtime with testing branch

rshar1 committed:
added back travis.yml

rshar1 committed:
Merge branch 'implementingRealtime' into realtime

rshar1 committed:
Merge remote-tracking branch 'origin/realtime' into realtime

turtlefarmer commented on src/codeu/chat/client/BroadCastReceiver.java:
```diff
@@ -0,0 +1,127 @@
+ */
+public class BroadCastReceiver extends Thread{
+
+    private ConnectionSource mySource;
```
source is probably a better variable name

turtlefarmer replied:
same for other variables.  member variables imply "my"

turtlefarmer commented on src/codeu/chat/client/BroadCastReceiver.java:
```diff
@@ -0,0 +1,127 @@
+    public void run() {
+
+        try (
+                final Connection myConnection = this.mySource.connect()
```
spacing is funky

turtlefarmer commented on src/codeu/chat/ServerMain.java:
```diff
@@ -51,21 +51,21 @@ public static void main(String[] args) {
 
       LOG.info("Starting server...");
-      runServer(id, secret, serverSource, relaySource);
+      runServer(id, secret, serverSource ,relaySource);
```
spacing

turtlefarmer commented on src/codeu/chat/ServerMain.java:
```diff
@@ -79,14 +79,18 @@ private static void runServer(Uuid id,
                                 ConnectionSource serverSource,
                                 ConnectionSource relaySource) {
 
+
```
I don't think this is a strict rule at Google, but most of the code I've seen have a "less is better" attitude towards vertical whitespaces, I assume for the sake of being able to see more code at a time on the screen.  

With that said, it's definitely okay to insert singular whitespaces to create logical segmentation.  Multiple vertical whitespaces is usually reserved for top-level separation though.

turtlefarmer commented on src/codeu/chat/SimpleGuiClientMain.java:
```diff
@@ -47,27 +48,30 @@ public static void main(String [] args) {
     ) {
-      final Controller controller = new Controller(source);
-      final View view = new View(source);
+      BroadCastReceiver receiver = new BroadCastReceiver(source);
```
final.

turtlefarmer commented on src/codeu/chat/server/BroadCastSystem.java:
```diff
@@ -0,0 +1,148 @@
+    private void removeConnection(Connection connection, Uuid uuid){
+
+        if (connection == null) throw new NullPointerException();
+        List<Connection> recipients = conversationUsers.get(uuid.id());
```
Micro optimization, but HashSet might be better performance for collections where you need to remove by object.

rshar1 replied:
Yes. I will switch it to HashSet. Thanks

turtlefarmer commented on src/codeu/chat/client/BroadCastReceiver.java:
```diff
@@ -0,0 +1,127 @@
+
+    }
+
+    public void joinConversation(ConversationSummary old, ConversationSummary newCon){
```
Why does joining operate on a ConversationSummary rather than a Conversation?

rshar1 replied:
To join a conversation, I only really care about the Uuid for the conversation they wish to join so I figured there was no need for the client and server to send an entire conversation instance

turtlefarmer commented on src/codeu/chat/client/BroadCastReceiver.java:
```diff
@@ -0,0 +1,127 @@
+                    int type = Serializers.INTEGER.read(in);
+
+                    if (type == NetworkCode.NEW_BROADCAST) {
+                        Message message = Message.SERIALIZER.read(in);
```
I'm not reviewing parts related to message serialization as carefully right now.  I'll take a closer look at the newer version.

turtlefarmer commented on src/codeu/chat/server/BroadCastSystem.java:
```diff
@@ -0,0 +1,148 @@
+
+    }
+
+    public void switchConversation(Connection connection, ConversationSummary oldCon, ConversationSummary newCon){
```
Similar to other comment, should these be operations on the Conversation object instead?

turtlefarmer commented on src/codeu/chat/server/BroadCastSystem.java:
```diff
@@ -0,0 +1,148 @@
+        if (oldCon != null)
+            removeConnection(connection, oldCon.id);
+        if (newCon != null)
+            addConnection(connection, newCon.id);
```
Is it still true that a user can only join one conversation at a time?  If so, does this code behave correctly if switchConvesration is called twice with null oldCon?

rshar1 replied:
Yes, this method will allow them to join more than one conversation at a time. The client side prevents this though. 

Would you suggest I add another check here? If I was to add such a check, I would use another HashMap that maps Connection to Conversation Uuid.

turtlefarmer commented on src/codeu/chat/server/BroadCastSystem.java:
```diff
@@ -0,0 +1,148 @@
+
+public class BroadCastSystem {
+
+    private class ConversationMessageLink {
```
Can we just store references to the conversation object instead of it's uuid?  If using a ConversationMessageLink object always requires us to convert from the deserialized UUid to the in-memory object anyways, we might as well do it during ConversationMessageLink creation.

rshar1 replied:
I could store the references to the conversation object but I'm not sure I understand what you mean in terms of the benefit. The reason I set it up this way is because the HashMap that keeps track of which users are in which conversations maps the Uuid to the Connections. As such, I never use the Conversation object..just its Uuid to find which connections are interested in these messages.

turtlefarmer commented on src/codeu/chat/server/BroadCastSystem.java:
```diff
@@ -0,0 +1,148 @@
+
+    public BroadCastSystem(){
+        conversationUsers = new ConcurrentHashMap<>();
+        messagesToBroadcast = new LinkedBlockingQueue<>();
```
Can types be specified for these collections

turtlefarmer commented on src/codeu/chat/server/BroadCastSystem.java:
```diff
@@ -0,0 +1,148 @@
+
+            try {
+                while (true) {
+                    broadCastMessage(messagesToBroadcast.take());
```
Consider:  Failure to broadcast right now causes the message to be lost.  Maybe considering re-inserting at end of the Queue so it can be retried.  If you do this, it'd be best to attach a retry count as well so messages that are doomed to fail don't retry indefinitely.

rshar1 replied:
So this exception is actually thrown when the blocking queue was interrupted while waiting for a message to be added. The issue is not with a specific message that failed to broadcast. 

If a failed attempt to broadcast a message happens, it will happen when sending to a specific user, so simply adding it back to queue would result in certain users getting the message more than once. 

I like the idea of adding multiple attempts to send a message because at the moment the connection is closed if there is a failed attempt to send a message. I will look into how to implement this.

turtlefarmer commented on src/codeu/chat/server/BroadCastSystem.java:
```diff
@@ -0,0 +1,148 @@
+
+    }
+
+    private ConcurrentHashMap<Integer, List<Connection>> conversationUsers;
```
Can the key just be the Conversation object?  Or maybe Conversation.UUID instead of Conversation.UUID.id()?
I'm not sure why they wrote their own UUID class, but from scanning it, i think HashCode is created in such a way that different objects with the same content will hash to the same thing.  If you can hash key by conversation, that'll make the rest of the code much more intuitive.

turtlefarmer replied:
Alternatively, does a mesasge not already belong to a conversation?  Can this just be a queue of messages?

rshar1 replied:
Currently, the messages do not store any information regarding which conversation they are a part of. Conversation only stores a reference to the first message. Each message then references the next message in the chain, similar to a linked list. 

In regards to the first comment, I'm thinking I will switch the Key to be Conversation.Uuid.

turtlefarmer commented on src/codeu/chat/server/BroadCastSystem.java:
```diff
@@ -0,0 +1,148 @@
+        if (conversationSummary == null) throw new NullPointerException("Cannot add null to conversations");
+        if (conversationSummary.id == null) throw new NullPointerException("Null ID in conversation summary");
+
+        int uid = conversationSummary.id.id();
```
I took another look at the provided UUID class.  If I'm understanding it correctly, it's just a chain of integers (since typically UUIDs are much bigger than int space).  I think your use of uuid.id() throughout the code is incorrect, since it's effectively just using the last int in the chain as a unique identifier.

rshar1 replied:
Yes. You are right. I will fix that up. Thanks

turtlefarmer commented on src/codeu/chat/server/BroadCastSystem.java:
```diff
@@ -0,0 +1,148 @@
+
+    // adds the given message to the list of messages that need to be broadcasted
+    public void addMessage(Uuid conversationUuid, Message message){
+        if (message == null) throw new NullPointerException("Message cannot be null");
```
maybe some logic for empty or trim()==empty messages as well.

rshar1 replied:
So if user decides they would like to send " " for whatever reason I shouldn't push the message? 

The broadcast system, pushes any new messages that have been accepted by the server. As such, if I add the logic here, the empty message will still appear in the model. Would it be more appropriate to add such logic to the server, rather than the broadcaster?

rshar1 committed:
fixed styling and spacing

Fixed the format of the BroadCastSystem and BroadCastReceiver.

Fixed some spacing issues and variable names.

rshar1 committed:
changed to hashes and fixed id issue

Changed the Collection keeping track of users in the broadcast system
from a linked list to a hashset.

Also changed the Type of the Hash map from integer to Uuid, since the
Integer was only keeping track of the last part of the Uuid.

rshar1 committed:
Merged google master into realtime

rshar1 committed:
Merge branch 'mergingGoogle' into mergingRealtime

rshar1 committed:
Merge branch 'serializerUpdates' into mergingSerializer

# Conflicts:
#	src/codeu/chat/client/View.java

rshar1 committed:
fixed view

rshar1 committed:
successfully merged serializer

rshar1 committed:
Merge branch 'sentiment-analysis' into mergingSentiment

# Conflicts:
#	src/codeu/chat/client/Controller.java
#	src/codeu/chat/client/View.java
#	src/codeu/chat/server/Server.java

rshar1 committed:
Merge branch 'sentiment-analysis' into mergingRealtime

# Conflicts:
#	run_client.sh
#	run_server.sh
#	test/codeu/chat/util/UuidTest.java

rshar1 committed:
fixed styling

Fixed the styling and removed some unnecessary imports

rshar1 committed:
slight changes to formatting

rshar1 committed:
Merge branch 'mergingRealtime' into implementingRealtime

rshar1 committed:
Merge branch 'implementingRealtime' into realtime

rshar1 committed:
changes to variable names

rshar1 committed:
added comment to client

Added comment to clients, showing something that should be changed

rshar1 committed:
uncommented log

Uncommented a server log entry that was previously commented.

rshar1 committed:
Merge branch 'testing' into realtime

# Conflicts:
#	run_client.sh
#	run_server.sh
#	src/codeu/chat/client/View.java
#	test/codeu/chat/util/UuidTest.java

turtlefarmer reviewed (approved):
I think this is fine.  In the future, try to separate formatting pull requests modifications from code change ones.  When they're intermingled it becomes really hard to see what the actual diff is.
